### PR TITLE
8231179: Investigate why tools/javac/options/BCPOrSystemNotSpecified.java fails on Window

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

The statement `String sourcePath = file.getName();` in method `prepareBCP(Path target)` would use `DirectoryFileObject::getName` which is shown below.

```
@Override @DefinedBy(Api.COMPILER)
public String getName() {
    return relativePath.resolveAgainst(userPackageRootDir).toString();  // <-----------------
}
```

And the `DirectoryFileObject::getName` would use `RelativePath::resolveAgainst` which is shown below

```
public Path resolveAgainst(Path directory) throws /*unchecked*/ InvalidPathException {
    String sep = directory.getFileSystem().getSeparator();  // <---------------------
    return directory.resolve(path.replace("/", sep));
}
```

We can see that `RelativePath::resolveAgainst` use file system separator instead of operating system separator. In method `prepareBCP`, the file system should be JRT file system, but we don't need to know actually which file system to be used. We only need to use the same file system separator so that the test can run as expected.

This patch fixes the test by using the file system separator instead of operating system separator.

Thank you for taking the time to review.

Best Regards.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8231179](https://bugs.openjdk.java.net/browse/JDK-8231179): Investigate why tools/javac/options/BCPOrSystemNotSpecified.java fails on Window

### Reviewers
 * [Vicente Romero](https://openjdk.java.net/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/2004/head:pull/2004` \
`$ git checkout pull/2004`

Update a local copy of the PR: \
`$ git checkout pull/2004` \
`$ git pull https://git.openjdk.java.net/jdk pull/2004/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2004`

View PR using the GUI difftool: \
`$ git pr show -t 2004`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/2004.diff">https://git.openjdk.java.net/jdk/pull/2004.diff</a>

</details>

### CODE REVIEW

File: test/langtools/tools/javac/options/BCPOrSystemNotSpecified.java [hunk 1]
```diff
@@ -206,7 +209,9 @@ private void prepareBCP(Path target) throws IOException {
                     Files.createDirectories(targetDir);
                     try (InputStream in = file.openInputStream()) {
                         String sourcePath = file.getName();
-                        int sepPos = sourcePath.lastIndexOf(fileSep);
+                        // Here, we should use file system separator instead of the operating system separator.
+                        String fileSystemSep = ((PathFileObject) file).getPath().getFileSystem().getSeparator();
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/options/BCPOrSystemNotSpecified.java:213]: what about using: System.getProperty("file.separator")?

Comment (jonathan-gibbons) [reply test/langtools/tools/javac/options/BCPOrSystemNotSpecified.java:213]: That would be incorrect.  
That system property is for the default file system, and may not be the correct separator for a user-defined file system, as is possible with `java.nio.file.FileSystem`

Comment (jonathan-gibbons) [inline test/langtools/tools/javac/options/BCPOrSystemNotSpecified.java:213]: It should be possible to get the `Path` without having to cast to the internal `PathFileObject` type.

See https://docs.oracle.com/en/java/javase/15/docs/api/java.compiler/javax/tools/StandardJavaFileManager.html#asPath(javax.tools.FileObject)

**Review by jonathan-gibbons — CHANGES_REQUESTED** [review]

File: test/langtools/tools/javac/options/BCPOrSystemNotSpecified.java (hunk 1 above)
Comment (jonathan-gibbons) [reply test/langtools/tools/javac/options/BCPOrSystemNotSpecified.java:213]: @vicente-romero-oracle Explaining a bit more ... on Windows, the default file system separator is `\` but we're dealing here with `Path` objects from the "internal" JRT file system, which uses `/` as a separator. That's why we have to be careful to use the right separator.  On Unix/Linux/macOS, the separators are the same, which is how the test has passed on those platforms.

Comment (lgxbslgx) [reply test/langtools/tools/javac/options/BCPOrSystemNotSpecified.java]: @jonathan-gibbons Thanks for your suggestion. I have updated the code.

Comment (vicente-romero-oracle) [reply test/langtools/tools/javac/options/BCPOrSystemNotSpecified.java]: I see thanks for the clarification

**Comment by lgxbslgx** [comment]
I find the windows-x64-build of the pre-submit-tests failed. I merge the code of the master branch to try to fix it.

**Comment by lgxbslgx** [comment]
I merged the code of the master branch just now to see whether the `tools/javac/options/BCPOrSystemNotSpecified.java` passes the `Pre-submit test`. If so, maybe it is time for us to push this patch.

**Comment by lgxbslgx** [comment]
I merged the master branch to solve the conflict. Thank you for reviewing.

**Review by vicente-romero-oracle — APPROVED** [review]
looks good to me

**Comment by lgxbslgx** [comment]
@vicente-romero-oracle Thank you for your review. Could I get your help to sponsor this patch?

/integrate

**Pull request closed without merging**